packages/clay-core/src/tree-view/useTree.ts

```diff
@@ -56,9 +79,148 @@ export function useTree(props: ITreeProps): ITreeState {
 		setExpandedKeys(expanded);
 	};
 
+	const open = (key: Key) => {
+		if (expandedKeys.has(key)) {
+			return;
+		}
+
+		const expanded = new Set(expandedKeys);
+
+		if (!expanded.has(key)) {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
 	return {
 		expandedKeys,
+		items,
+		open,
+		reorder,
 		selection,
 		toggle,
 	};
 }
+
+// RFC 6902 4.4
+type Patch = {
+	op: 'move';
+	from: Array<number>;
+	path: Array<number>;
+};
+
+export function createImmutableTree<T extends Array<Record<string, any>>>(
+	initialTree: T,
+	nestedKey: string
+) {
+	const patches: Array<Patch> = [];
+
+	let immutableTree = [...initialTree] as T;
+
+	function pointer(tree: T, index: number, item: any) {
+		return [...tree.slice(0, index), item, ...tree.slice(index + 1)] as T;
+	}
+
+	function nodeByPath(path: Array<number>) {
+		const queue = [...path];
+		const rootIndex: number = queue.shift() as number;
+
+		let item = {...immutableTree[rootIndex]};
+		let parent = null;
+		let index = rootIndex;
+
+		immutableTree = pointer(immutableTree, index, item);
+
+		while (queue.length) {
+			index = queue.shift() as number;
+
+			if (Array.isArray(item[nestedKey]) && item[nestedKey].length) {
+				parent = item;
+				item = {...item[nestedKey][index]};
+
+				parent[nestedKey] = pointer(parent[nestedKey], index, item);
+			}
+		}
+
+		return {
+			index,
+			item,
+			parent,
+		};
+	}
+
+	function applyPatches(): T {
+		patches.forEach((patch) => {
+			const {from, op, path} = patch;
+
+			switch (op) {
+				// Applies the operation on the tree, the move is functionally
+				// identical to a "remove" operation on the `from` location and
+				// immediately followed by the "add" operation at the target
+				// location with the value that was removed.
+				case 'move': {
+					const nodeToRemove = nodeByPath(from);
+
+					if (nodeToRemove.parent) {
+						nodeToRemove.parent[nestedKey] = nodeToRemove.parent[
+							nestedKey
+						].filter(
+							(item: any, index: number) =>
+								index !== nodeToRemove.index
+						);
+					} else {
+						immutableTree = immutableTree.filter(
+							(item: any, index: number) =>
+								index !== nodeToRemove.index
+						) as T;
+					}
+
+					const pathToAdd = nodeByPath(path);
+
+					if (pathToAdd.parent) {
+						if (!Array.isArray(pathToAdd.parent[nestedKey])) {
+							pathToAdd.parent[nestedKey] = [];
+						}
+
+						pathToAdd.parent[nestedKey].splice(pathToAdd.index, 1);
+						pathToAdd.parent[nestedKey].splice(
+							pathToAdd.index,
+							0,
+							nodeToRemove.item
+						);
+						pathToAdd.parent[nestedKey].splice(
+							pathToAdd.index + 1,
+							0,
+							pathToAdd.item
+						);
+					} else {
+						if (!Array.isArray(pathToAdd.item[nestedKey])) {
+							pathToAdd.item[nestedKey] = [];
+						}
+
+						pathToAdd.item[nestedKey] = [
+							...pathToAdd.item[nestedKey],
+							nodeToRemove.item,
+						];
+					}
```

Comment:
I've tested this use case a bit and it looks good but it's still a little weird when you try to move an item inside an item, it will add a sibling to the item instead of adding inside, I think we'll have to do something similar that we did in the Page Editor, I had made a calculation that 20% of the item's edge, will move up when it's outside of that 20%, it will move inward, the same is done for the item's bottom edge, this can be done before calling the `reorder` function.

But I think we would still have to use a different behavior and not use `parent`, maybe we have to determine to use the parent just in case when we move up and down and use the item `item[nestedKey]` to add the item inside, the problem is that we will almost always have parent, maybe comparing the size of the `from` and `path` along with the index values, we can know if the user wants to move inside the item.

Reply:
> I've tested this use case a bit and it looks good but it's still a little weird when you try to move an item inside an item, it will add a sibling to the item instead of adding inside, I think we'll have to do something similar that we did in the Page Editor, I had made a calculation that 20% of the item's edge, will move up when it's outside of that 20%, it will move inward, the same is done for the item's bottom edge, this can be done before calling the reorder function.

Yes, I agree and I'll have a look into that.

> But I think we would still have to use a different behavior and not use parent

Yes, I actually removed that in https://github.com/liferay/clay/commit/a118a2aceb31663fc1f06dc822b1754819f5973f, Now we also need to add the proper styles, looking at the scss files, I think we're going to need @pat270 's help for that.